### ISSUE TITLE
sy.BaseDataset.send() incurs extraordinary MEMORY USAGE

### ISSUE THREAD

**Issue by wentaiwu92** [OpenMined/PySyft#2499]
**Description**
Invoking sy.BaseDataset.send(some_worker) incurs memory consumption roughly equivalent to the number of workers times the size of this part of data.

**To Reproduce**
```
# split data, it is Okay, no extra memory usage
for i in range(n_clients):
    # prepare client data, train and test separately
    client_train_data.append(
        sy.BaseDataset(data_train_x[split_points_train[i]: split_points_train[i+1]],
                       data_train_y[split_points_train[i]: split_points_train[i+1]]))  # non-overlapping data fragment
    client_test_data.append(
        sy.BaseDataset(data_test_x[split_points_test[i]: split_points_test[i+1]],
                       data_test_y[split_points_test[i]: split_points_test[i+1]]))  # non-overlapping data fragment

# now MEM usage is still roughly equal to size of my data (100MB)

# dispatching - where the problem is
for i in range(n_clients):
    client_train_data[i].send(clients[i])
    client_test_data[i].send(clients[i])

# now MEM usage reaches 10GB!
...
```

**Expected behavior**
Assume size(client_train_data+client_test_data) = 100MB, n_clients = 100. After invoking send() for all clients (as the codes above), memory usage increases from 100~200MB to 10GB, which is 100MB * 100.

If I changed n_clients to 10 and 200, the ultimate memory usage becomes 1GB and 20GB, respectively. Results of repeated trials proved my suspicion on the problem of **send()**

**Screenshots**
with n_clients = 100, size of my dataset roughly equals to 100MB:
![image](https://user-images.githubusercontent.com/46561082/62933172-82a59580-bdb9-11e9-816c-5ab33cca7163.png)

**Desktop (please complete the following information):**
 - OS: win10
 - Version Python 3.7

**Additional context**
You have to set a big-enough dataset and a sufficient number of clients, otherwise the problem is not easy to reveal.

**Comment by wentaiwu92** [comment]
**Further findings**
With the suspicion that not only the sliced data fragment but also the whole data set are sent to clients when send() is invoked, I tried to detach the tensors (using .detach() or .data) before wrapping them into the **_BaseDataset_** objects.  However, PySyft yielded an error indicating that detached tensor is not eligible. To avoid direct detach() operation, I further tried to use clone() plus requires_grad_(False). As a result, memory usage **_declines_** to a relatively normal level.

**Probable cause**
I guess the very likely cause of this problem is with the computation graph, which may be sent to every client together with their sliced data. This also leads to the copy of the **_entire dataset_** because the dataset (as a tensor) is the predecessor of every slice of data in the graph (and PySyft may need to keep it for back propagation), or maybe all the slices share the same node as the original dataset - I'm not an expert on computation graph of Tensors.

So a possible solution is to create separate tensors for each BaseDataset/client in FL with requires_grad_=False (avoid back-propagation and the corresponding link in the graph). At least simply slicing the original data is not feasible though it's the most intuitive implementation.

**Possible solution**
```
# split data and dispatch
    for i in range(env_cfg.n_clients):
        # prepare client data, train and test separately
        client_train_data.append(
            sy.BaseDataset(data_train_x[split_points_train[i]: split_points_train[i+1]].clone().requires_grad_(False),
                           data_train_y[split_points_train[i]: split_points_train[i+1]].clone().requires_grad_(False)))
        client_test_data.append(
            sy.BaseDataset(data_test_x[split_points_test[i]: split_points_test[i+1]].clone().requires_grad_(False),
                           data_test_y[split_points_test[i]: split_points_test[i+1]].clone().requires_grad_(False)))
```

**Comment by mari-linhares** [comment]
@kamathhrishi
